cmake/SetupDoxygen.cmake

```diff
@@ -14,25 +14,64 @@ if (DOXYGEN_FOUND)
   # Configure file that contains doxygen groups
   configure_file(docs/GroupDefs.hpp ${SPECTRE_DOXYGEN_GROUPS})
 
-  add_custom_target(
-      doc
-      COMMAND ${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml 2>&1
-      DEPENDS
-      ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml
-      ${SPECTRE_DOXYGEN_GROUPS}
-      )
-
-  add_custom_target(
-    doc-check
-    # search for warnings and print out 6 lines after to make sure
-    # majority of warning is printed so the results are actually
-    # useful.
-    COMMAND ! ${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml 2>&1
-    | grep -A 6 'warning'
-    DEPENDS
-    ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml
-    ${SPECTRE_DOXYGEN_GROUPS}
-    )
+  # We need Python for postprocessing the documentation
+  find_package(PythonInterp)
+  include(FindPythonModule)
+  find_python_module(bs4 FALSE)
+  find_python_module(pybtex FALSE)
+
+  if (PYTHONINTERP_FOUND AND PY_BS4 AND PY_PYBTEX)
+    add_custom_target(
+        doc
+        COMMAND
+        ${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml 2>&1
+        && ${PYTHON_EXECUTABLE}
+        ${CMAKE_SOURCE_DIR}/docs/config/postprocess_docs.py -v
+        --html-dir ${PROJECT_BINARY_DIR}/docs/html
+        --references-file ${CMAKE_SOURCE_DIR}/docs/References.bib
+        DEPENDS
+        ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml
+        ${SPECTRE_DOXYGEN_GROUPS}
+        )
+    add_custom_target(
+        doc-check
+        # search for warnings and print out 6 lines after to make sure
+        # majority of warning is printed so the results are actually
+        # useful.
+        COMMAND
+        ! ${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml 2>&1
+        | grep -A 6 'warning'
+        && ${PYTHON_EXECUTABLE}
```

Comment:
`tools/pr-docs` (and my personal script it was originally based on) depend upon `make doc-check` creating the documentation even if there were warnings.  But since this just skips the post-processing maybe it's OK.

Reply:
The post-processing does not interfere with that. It just runs over what Doxygen has generated.

Reply:
But it doesn't run if there are warnings was what I was getting at.

Reply:
Right, the postprocessing is conditioned on Doxygen succeeding without warnings. That's also what we test for in Travis. It means that while there are warnings there are things to fix in the PR.

Reply:
Yes, but I don't think that's the desired behavior.  This target is not solely for the purpose of checking for warnings.  I use it the same was as the C++ make targets: The primary purpose is to generate the output, but as a side effect it reports warnings that will need to be fixed at some point.

Reply:
Since we treat warnings as errors, it fails at the first error. I think that’s a useful behavior.

If you‘d like it to behave differently, how would you like it changed?

Reply:
I don't know whether there are restrictions from CMake, but I think the cleanest way would be something like
```sh
doxygen-thing | grep-thing
ret=$?
postprocess-thing && exit ${ret}
```
That is, save the result of the grep and exit with it if the post-process succeeds.  If the post-process fails this continues past the last line, but if that's the end of the script that should be fine (and will return failure because the last command did).

Reply:
Okay, that's worth trying. CMake does allow some reasonable amount of shell things, [here](https://github.com/sxs-collaboration/spectre/blob/8839a2633b0289f4dc816c651c40479cbc15fa6c/cmake/CodeCoverage.cmake#L85) is an example in our code already. I guess worst case the return value could be printed into a file and read in a later command? Since generating documentation already does a bunch of IO I don't think one extra small file will make any difference :)

Reply:
the problem is cmake, where I haven't found a way to store an intermediate shell variable like your `ret=$?`. I have spent quite a bit of time on this now but wouldn't like to waist much more.

Reply:
I don't know how to do this other than adding a small script to `docs/config` that is then run by the target. @wthrowe any other ideas?

Reply:
We do that already with many things using CMake's `configure_file` method, so that won't be a limitation. It just means that a copy of the script will be placed in `BUILD_DIR/bin` (or something like that, maybe `BUILD_DIR/tmp`)

Reply:
Ok I found a way to do this by writing the shell script into a file. Please check the fixup.